### PR TITLE
Add support for APIv2 Query format

### PR DESCRIPTION
**Pull request by JAVGan**
# Add support for APIv2 Query format

This PR implements the new StArMap `APIv2` responses while it still keeping a temporary support for APIv1 with some backwards incompatible changes.

The `APIv1` format may be then marked as deprecated and removed in future iterations, at this point I prefer to keep it around for this PR.

# Changes

The PR includes 3 commits, being the first one responsible for the majority of changes and additional testing files.

The commits are:
1. Add models for StArMap APIv2 Query
2. Implement support for APIv2 on Client
3. Add support for APIv2 offline mappings

It's recommended to go commit over commit to understand the proposed changes from this PR.

## Add models for StArMap APIv2 Query 

 This commit introduces the following new models for the StArMap APIv2:
 
- `QueryResponseContainer`: holds the whole Query APIv2 response within its property `responses` which is a list of `QueryResponseEntity`.

- `QueryResponseEntity`: Represents a single mapping for a given workflow, name and cloud. It can be converted to the `QueryResponse` from APIv1 model which works in an analogue way (without support for cloud name as a property).

- `MappingResponseObject`: Represent a single mapping entity which is part of `QueryResponseEntity`. It holds a list of `Destination` object, similar to what `QueryResponse` does on its `clouds` attribute.

- `BillingCodeRule`: A dedicated model to represent a Billing Code Rule for the community workflow.

It also adds a lot of JSON testing files to validate the new models behavior and ensure they're properly converting the expected data.

## Implement support for APIv2 on Client
    
This commit changes the `StarmapClient` to be able to handle the APIv2 calls to retrieve a `QueryResponseContainer` whenever `v2` endpoints are invoked.
    
It also changes the default `api_version` to `v2` and update the tests.
    
## Add support for APIv2 offline mappings
    
This commit updates the `providers.memory` module with the following changes:
    
- Rename `InMemoryMapProvider` to `InMemoryMapProviderV1` which will  only be used by APIv1 responses.
- Create a new provider named `InMemoryMapProviderV2` for APIv2 responses.
    
It also updates the provider's typing annotations to support a more generic approach on responses format.

# Backward incompatible changes

1. The offline provider `InMemoryMapProvider` was renamed to `InMemoryMapProviderV1`
2. The library no longers enforce the `stratosphere` workflow by default for APIv1. All workflows must be explicitly passed by the query methods. This, however, doesn't change the default behavior of the server.
3. The return typing annotation for query methods now consider the possibility to return `QueryResponseContainer` objects. The library user must properly check which object has been returned from queries, being either: `QueryResponse` or `QueryResponseContainer`.

# JIRA
Refers to SPSTRAT-382

### CODE REVIEW

**Comment by JAVGan** [comment]
@lslebodn @jajreidy PTAL

File: starmap_client/utils.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    Args:
+        a (dict):
+            The left dictionary to be combined
+        b (dict):
+            The right dictionary to be combined. It will override the same keys from A.
+    Returns:
+        dict: A new dictionary with combination of A and B.
+    """
+    c = {}
```
Comment (jajreidy) [inline starmap_client/utils.py:21]: instead of c = {} you could do 'return a | b' . Just a suggestion

Comment (JAVGan) [reply starmap_client/utils.py:21]: I've tried that but IIUC `a | b` requires the keys to not overlap.

In this method we're considering that one or more keys may be overlapping so `b` would be determinant over `a` if that happens.

Comment (JAVGan) [reply starmap_client/utils.py:21]: Ok, let me try to adjust it 😄

Comment (JAVGan) [reply starmap_client/utils.py]: I've changed it, thanks for the suggestion!

Comment (JAVGan) [reply starmap_client/utils.py]: Oh.. It seems to have broken `py38` 😞

Comment (JAVGan) [reply starmap_client/utils.py]: I'm not sure whether we can drop it since `pub` might still use this version...

@lslebodn thoughts?

Comment (JAVGan) [reply starmap_client/utils.py]: @lslebodn JFTR: https://github.com/release-engineering/starmap-client/compare/b892f06ab1d4c91106206de582d4051fd6cb5284..c8c6ff4805a15e4b0491290a1bc1e52b284e76f0

Comment (jajreidy) [reply starmap_client/utils.py]: I'm sorry! I forgot that pub still has py38. From what I've read the | operator was added py39+

Comment (JAVGan) [reply starmap_client/utils.py]: @jajreidy Added on 3d49924361e696485e8b7226cde895886b69def4

@lslebodn thanks a lot for allowing me to merge #47

**Review by jajreidy — APPROVED** [review]
LGTM

**Pull request merged**